### PR TITLE
Add mappings for enrich fields

### PR DESCRIPTION
**Pull request by dnhatn**
We are developing a new component that enriches the search results during query time. The idea is to leverage the enrich policies and indices already used during indexing time. However, for the new component to function properly, we require the mapping types and doc_values of the enrich fields.

~Currently, an enrich index can have multiple source indices, and any mapping conflicts that arise between the enrich fields of these source indices are ignored. However, with the upcoming changes, the enrich policy execution will fail if any mapping conflicts occur. We consider this change not as a breaking change, but rather as addressing a validation bug, as enrich fields should be consistent across source indices.~

### CODE REVIEW

**Review by nik9000 — COMMENTED** [review]
The tests sure look like what we want.

**Comment by dnhatn** [comment]
@jbaiera @martijnvg I have updated this PR to maintain the current behavior. Populating the mapping for enrich fields is now a best-effort, which means we will ignore mapping conflicts that arise between the enrich fields of these source indices. I will have a follow-up to adjust this validation. Would you please take a look? Thank you!

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java [hunk 1]
```diff
@@ -2124,4 +2301,11 @@ private void ensureEnrichIndexIsReadOnly(String createdEnrichIndex) {
 
         assertThat(expected.getMessage(), containsString("index [" + createdEnrichIndex + "] blocked by: [FORBIDDEN/8/index write (api)]"));
     }
+
+    private static void assertEnrichMapping(Map<String, Object> actual, String expectedMapping) {
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:2305]: :+1:

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 2]
```diff
@@ -194,11 +195,56 @@ static void validateMappings(
...
+                    Strings.collectionToCommaDelimitedString(policy.getIndices()),
+                    Strings.collectionToCommaDelimitedString(formats)
+                );
+            }
+            return null;
+        }
+        return new MappingTypeAndFormat(Iterables.get(types, 0), formats.isEmpty() ? null : Iterables.get(formats, 0));
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:242]: is it possible to not not suppress unchecked warnings?

**Review by martijnvg — APPROVED** [review]
LGTM

I think having the enrich fields mapped is a good change in general.
Enabling doc values by default does increase the size of the enrich index.
This shouldn't be too bad, given that an enrich index is designed to contain reference data set for lookups, an enrich index usually isn't that large. And it allows enrich to be used at query time for join like functionality. Down the line enrich indices could be configured to not store _source or the  synthesise _source at runtime.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:242]: ++, I pushed https://github.com/elastic/elasticsearch/pull/96056/commits/db42ed0dc2c37006ca9424920790845edb30a9d7 to reduce the scope of the suppression.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 3]
```diff
@@ -194,11 +195,53 @@ static void validateMappings(
         }
     }
 
-    private static void validateField(Map<?, ?> properties, String fieldName, boolean fieldRequired) {
+    private record MappingTypeAndFormat(String type, String format) {
+
+    }
+
+    private static MappingTypeAndFormat getMappingTypeAndFormat(
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: I renamed `getMappingTypeAndFormat` -> `validateAndGetMappingTypeAndFormat` in https://github.com/elastic/elasticsearch/pull/96056/commits/4e9656c895fa8cdf6ee33844dc25a5e7b49b520f.

**Comment by dnhatn** [comment]
@nik9000 @martijnvg Thank you for reviews.

**Pull request merged**